File: android/src/main/java/app/devlas/plugins/thermalprinter/TcpTransport.kt
package app.devlas.plugins.thermalprinter

import java.io.IOException
import java.net.InetSocketAddress
import java.net.Socket
import java.net.SocketTimeoutException

/**
 * Transporte TCP — impresión raw a puerto 9100 (JetDirect), el estándar de
 * térmicas de red cableada y WiFi. Sin librerías: un Socket y ya.
 */
class TcpTransport {

    companion object {
        private const val CONNECT_TIMEOUT_MS = 4000
        private const val CHUNK_SIZE = 8192
        /** Corto a propósito: una impresora que no soporta DLE EOT nunca responde. */
        private const val STATUS_READ_TIMEOUT_MS = 800
    }

    /**
     * Pregunta las cuatro consultas de estado y devuelve los bytes crudos.
     * `null` en una posición significa "no contestó" — ver PrinterStatus.
     */
    fun status(host: String, port: Int): List<Int?> {
        val selectors = listOf(
            PrinterStatus.N_PRINTER,
            PrinterStatus.N_OFFLINE,
            PrinterStatus.N_ERROR,
            PrinterStatus.N_PAPER,
        )
        try {
            Socket().use { socket ->
                socket.connect(InetSocketAddress(host, port), CONNECT_TIMEOUT_MS)
                socket.soTimeout = STATUS_READ_TIMEOUT_MS
                val out = socket.getOutputStream()
                val input = socket.getInputStream()

                return selectors.map { n ->
                    try {
                        out.write(PrinterStatus.query(n))
                        out.flush()
                        val b = input.read()
                        if (b < 0) null else b
                    } catch (e: SocketTimeoutException) {
                        null
                    }
                }
            }
        } catch (e: SocketTimeoutException) {
            throw PrinterException("connect_failed", "Timeout conectando a $host:$port")
        } catch (e: IOException) {
            throw PrinterException("connect_failed", "No se pudo consultar $host:$port: ${e.message}")
        }
    }

    fun print(host: String, port: Int, bytes: ByteArray) {
        try {
            Socket().use { socket ->
                socket.connect(InetSocketAddress(host, port), CONNECT_TIMEOUT_MS)
                val out = socket.getOutputStream()
                var offset = 0
                while (offset < bytes.size) {
                    val len = minOf(CHUNK_SIZE, bytes.size - offset)
                    out.write(bytes, offset, len)
                    offset += len
                }
                out.flush()
                // Espera proporcional antes de cerrar — evita truncar el búfer de la
                // impresora en trabajos grandes (patrón DantSu: ~1ms por 16 bytes).
                Thread.sleep(minOf(2000L, bytes.size / 16L))
            }
        } catch (e: SocketTimeoutException) {
            throw PrinterException("connect_failed", "Timeout conectando a $host:$port")
        } catch (e: IOException) {
            throw PrinterException("connect_failed", "No se pudo imprimir en $host:$port: ${e.message}")
        }
    }
}

File: ios/Sources/ThermalPrinterPlugin/TcpTransport.swift
import Foundation
import Network

/// Transporte TCP — impresión raw a puerto 9100 (JetDirect), el estándar de
/// térmicas de red cableada y WiFi. Usa Network.framework (NWConnection) —
/// nativo de iOS, sin librerías externas.
///
/// Requiere `NSLocalNetworkUsageDescription` en el Info.plist de la app host
/// (iOS pide permiso de "red local" la primera vez que se conecta a una IP
/// privada — sin ese string el sistema rechaza el diálogo silenciosamente).
final class TcpTransport {

    private static let connectTimeoutMs = 4000

    func print(host: String, port: UInt16, bytes: Data, completion: @escaping (Result<Void, PrinterError>) -> Void) {
        guard let nwPort = NWEndpoint.Port(rawValue: port) else {
            completion(.failure(PrinterError("connect_failed", "Puerto inválido: \(port)")))
            return
        }

        let connection = NWConnection(host: NWEndpoint.Host(host), port: nwPort, using: .tcp)
        let queue = DispatchQueue(label: "cl.devlas.thermalprinter.tcp")
        var finished = false

        func finish(_ result: Result<Void, PrinterError>) {
            guard !finished else { return }
            finished = true
            connection.cancel()
            completion(result)
        }

        let timeoutWork = DispatchWorkItem {
            finish(.failure(PrinterError("connect_failed", "Timeout conectando a \(host):\(port)")))
        }
        queue.asyncAfter(deadline: .now() + .milliseconds(Self.connectTimeoutMs), execute: timeoutWork)

        connection.stateUpdateHandler = { state in
            switch state {
            case .ready:
                timeoutWork.cancel()
                connection.send(content: bytes, completion: .contentProcessed { error in
                    if let error = error {
                        finish(.failure(PrinterError("connect_failed", "No se pudo imprimir en \(host):\(port): \(error.localizedDescription)")))
                        return
                    }
                    // Espera proporcional antes de cerrar — evita truncar el búfer de la
                    // impresora en trabajos grandes (patrón DantSu: ~1ms por 16 bytes).
                    let delayMs = min(2000, bytes.count / 16)
                    queue.asyncAfter(deadline: .now() + .milliseconds(delayMs)) {
                        finish(.success(()))
                    }
                })
            case .failed(let error):
                timeoutWork.cancel()
                finish(.failure(PrinterError("connect_failed", "No se pudo conectar a \(host):\(port): \(error.localizedDescription)")))
            case .cancelled:
                timeoutWork.cancel()
            default:
                break
            }
        }
        connection.start(queue: queue)
    }
}
